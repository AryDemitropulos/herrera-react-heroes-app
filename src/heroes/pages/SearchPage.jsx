import queryString from 'query-string';

import { useLocation, useNavigate } from 'react-router';
import { useForm } from '../../hooks';
import { HeroCard } from '../components/HeroCard';
import { getHeroesByName } from '../helpers/getHeroByName';

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  console.log(location);
  const { q: query = '' } = queryString.parse(location.search);

  const { searchText, onInputChange } = useForm({ searchText: query });

  const heroes = getHeroesByName(query);

  const onSearchSubmit = (event) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
  };

  const hero = null;
  return (
    <>
      <h1>SearchPage</h1>
      <hr />
      <div className='row'>
        <div className='col-5'>
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit} aria-label='form'>
            <input
              type='text'
              placeholder='Search a hero'
              className='form-control'
              name='searchText'
              autoComplete='off'
              value={searchText}
              onChange={onInputChange}
            />
            <button className='btn btn-outline-primary mt-3'>Search</button>
          </form>
        </div>

        <div className='col-7'>
          <h4>Results</h4>
          <hr />

          {!query && (
            <div aria-label='alert-default' className='alert alert-primary'>
              Search a hero
            </div>
          )}

          {query && heroes.length === 0 && (
            <div aria-label='alert-danger' className='alert alert-danger'>
              No hero with <b>{query}</b>
            </div>
          )}

          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
