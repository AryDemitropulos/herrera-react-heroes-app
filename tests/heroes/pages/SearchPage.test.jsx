import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockedNavigate = jest.fn();

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useNavigate: () => mockedNavigate,
}));

describe('Prubas sobre <SearchPage/>', () => {
  beforeEach(() => jest.clearAllMocks());

  test('debe de mostrarse por con valores por defecto', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar a batman y el input con valor del queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');

    const img = screen.getByRole('img');
    expect(img.src).toBe('http://localhost/heroes/dc-batman.jpg');
  });

  test('debe de mostrar alert "Search a Hero" si no buscÃ³ ningun heroe', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    );
    const defaulAlert = screen.getByLabelText('alert-default');
    const dangerAlert = screen.queryByLabelText('alert-danger');
    expect(dangerAlert).toBeNull();

    expect(defaulAlert).toBeTruthy();
  });

  test('debe de llamar al navigate a la pantalla nueva', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    );
    const form = screen.getByLabelText('form');
    const input = screen.getByRole('textbox');

    fireEvent.change(input, { target: { value: 'testSearch' } });
    fireEvent.submit(form);

    expect(mockedNavigate).toHaveBeenCalledWith('?q=testSearch');
  });
});
