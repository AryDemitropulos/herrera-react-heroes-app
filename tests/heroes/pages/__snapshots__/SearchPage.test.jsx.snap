// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prubas sobre <SearchPage/> debe de mostrarse por con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alert-default"
        class="alert alert-primary"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
